src/common/base64.c

```diff
@@ -31,3 +32,101 @@ int base64_encode(const uint8_t *data, size_t data_length, char *out, size_t out
+    15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 63,  //
+    64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,  //
+    41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,  //
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64   //
+};
+
+// Taken from Exchange application implementation
+int base64url_decode(const char *src, size_t src_len, uint8_t *dest, size_t dest_len) {
+    const uint8_t *original_dest;
+
+    original_dest = dest;
+
+    if (src_len % 4 == 1) {
+        return -1;
+    } else if (src_len == 0) {
+        return 0;
+    }
+
+    while (src_len > 4) {
+        if (dest_len < 3) {
+            return -1;
+        }
+        *(dest++) = (pr2six[(uint8_t) src[0]] << 2 | pr2six[(uint8_t) src[1]] >> 4);
+        *(dest++) = (pr2six[(uint8_t) src[1]] << 4 | pr2six[(uint8_t) src[2]] >> 2);
+        *(dest++) = (pr2six[(uint8_t) src[2]] << 6 | pr2six[(uint8_t) src[3]] >> 0);
+        src += 4;
+        src_len -= 4;
+        dest_len -= 3;
+    }
+
+    if (src_len > 1) {
+        if (dest_len < 1) {
+            return -1;
+        }
+        *(dest++) = (pr2six[(uint8_t) src[0]] << 2 | pr2six[(uint8_t) src[1]] >> 4);
+        dest_len--;
+    }
+
+    if (src_len > 2) {
+        if (dest_len < 1) {
+            return -1;
+        }
+        *(dest++) = (pr2six[(uint8_t) src[1]] << 4 | pr2six[(uint8_t) src[2]] >> 2);
+        dest_len--;
+    }
+
+    if (src_len > 3) {
+        if (dest_len < 1) {
+            return -1;
+        }
+        *(dest++) = (pr2six[(uint8_t) src[2]] << 6 | pr2six[(uint8_t) src[3]] >> 0);
+        dest_len--;
+    }
+
+    return dest - original_dest;
+}
+
+int base64_to_base64url(const char *src, size_t src_len, char *dest, size_t dest_len) {
+    if (dest_len < src_len) {
+        return -1;
+    }
+    explicit_bzero(dest, dest_len);
+
+    size_t i;
+    for (i = 0; i < src_len; i++) {
+        switch (src[i]) {
+            case '+':
+                dest[i] = '-';
+                break;
+            case '/':
+                dest[i] = '_';
+                break;
+            case '=':
+                // Padding character only found at the end, we simply remove it
+                dest[i] = '\0';
+                return i;
+            default:
+                dest[i] = src[i];  // Copy other characters directly
+                break;
+        }
+    }
+
+    return src_len;
+}
```

Comment:
should be added in the SDK ?

Reply:
:+1: This was also my feeling

Reply:
could be handle as part as the "swap SDK" improvement